[Clang] [C++26] Expansion Statements (Part 9: Control Flow)

Sirraide opened the pull request:
This disallows certain `switch`-`case` constructs and labels in expansion statements.

GNU local labels are supported as an extension. My reasoning for allowing this is that:
1. I’d argue it makes sense: the main reason for disallowing labels is that you’d end up w/ more than one label w/ the same name in a single function, and jumping to such a label from outside an expansion statement becomes rather ill-defined as a result (which instantiation does it jump to? what if the statement expands to nothing at all?). Both of these issues don’t exist with local labels since they aren’t scoped to the entire function and as a result can only be jumped to from inside their respective instantiation.
2. Codegen and constant evaluation for this basically just works.
3. Disallowing this in the most straight-forward manner (e.g. by just dropping the label) leads to rather poor QOI and a lot of diagnostics, so even if we wanted to disallow local labels... we’d basically still have to do everything I’m doing in this patch in addition to then issuing a diagnostic for good QOI.

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 9)

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 9)

Sirraide commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -3738,6 +3738,12 @@ def err_expansion_stmt_incomplete : Error<
+def err_expansion_stmt_case : Error<
+  "%select{'case'|'default'}0 belongs to 'switch' outside enclosing expansion statement">;
+def note_enclosing_switch_statement_here : Note<
+  "switch statement is here">;
```
Apparently we really don’t

Sirraide commented on clang/test/SemaCXX/cxx2c-expansion-stmts-control-flow.cpp:
```diff
@@ -0,0 +1,117 @@
+  template for (auto x : {1, 2}) {
+    switch (i) {
+      case 1:;
+      default:
```
> Is it legal to omit the ";" after a "default:"?

I believe that was made legal in C++20 or C++23, I don’t remember exactly what version, but it is legal now.

cor3ntin commented on clang/lib/Parse/ParseStmt.cpp:
```diff
@@ -751,8 +751,9 @@ StmtResult Parser::ParseLabeledStatement(ParsedAttributes &Attrs,
-                                              IdentTok.getLocation());
+  LabelDecl *LD = Actions.LookupOrCreateLabel(
+      IdentTok.getIdentifierInfo(), IdentTok.getLocation(), /*GnuLabelLoc=*/{},
+      /*ForLabelStmt=*/true);
```
Am i missing something or is this always true?

Sirraide replied:
It has a default parameter of `false`; irrc it is false when we’re parsing the label of a `goto` rather than a `label:` declaration

cor3ntin commented on clang/lib/Sema/SemaLookup.cpp:
```diff
@@ -4468,7 +4468,8 @@ LabelDecl *Sema::LookupExistingLabel(IdentifierInfo *II, SourceLocation Loc) {
 LabelDecl *Sema::LookupOrCreateLabel(IdentifierInfo *II, SourceLocation Loc,
-                                     SourceLocation GnuLabelLoc) {
+                                     SourceLocation GnuLabelLoc,
+                                     bool ForLabelStmt) {
```
```suggestion
                                     bool IsLabelStmt) {
```

cor3ntin commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -9612,7 +9612,8 @@ class Sema final : public SemaBase {
   LabelDecl *LookupOrCreateLabel(IdentifierInfo *II, SourceLocation IdentLoc,
-                                 SourceLocation GnuLabelLoc = SourceLocation());
+                                 SourceLocation GnuLabelLoc = SourceLocation(),
+                                 bool ForLabelStmt = false);
```
It would be useful to add documentation for the new param

shafik commented on clang/include/clang/Sema/ScopeInfo.h:
```diff
@@ -202,7 +202,11 @@ class FunctionScopeInfo {
 public:
   /// A SwitchStmt, along with a flag indicating if its list of case statements
   /// is incomplete (because we dropped an invalid one while parsing).
```
Shouldn't this comment be updated to document the new data.

shafik commented on clang/lib/Parse/ParseStmt.cpp:
```diff
@@ -783,6 +784,12 @@ StmtResult Parser::ParseLabeledStatement(ParsedAttributes &Attrs,
 
   DiagnoseLabelFollowedByDecl(*this, SubStmt.get());
 
+  // If a label cannot appear here, just return the underlying statement.
```
I would add to the comment noting it will be diagnosed in `LookupOrCreateLabel` already.

shafik commented on clang/lib/Sema/SemaStmt.cpp:
```diff
@@ -1196,8 +1221,9 @@ StmtResult Sema::ActOnStartOfSwitchStmt(SourceLocation SwitchLoc,
 
   auto *SS = SwitchStmt::Create(Context, InitStmt, Cond.get().first, CondExpr,
                                 LParenLoc, RParenLoc);
+  SS->setSwitchLoc(SwitchLoc);
```
Why move the setting of the switch location here?

Sirraide replied:
We want to be able to point to the switch statement in a diagnostic before we’re done parsing the body, so I moved setting the location here rather than doing that as part of the call to `SS->setBody()` in `ActOnFinishSwitchStmt()`

shafik commented on clang/test/SemaCXX/cxx2c-expansion-stmts-control-flow.cpp:
```diff
@@ -0,0 +1,135 @@
+  }
+
+  // GNU local labels are allowed.
+  template for (auto x : {1, 2}) {
```
Looks like gcc does not support this: https://godbolt.org/z/Pz7sdq1Ge

Maybe we want to check w/ them if this was an oversight or not.

Sirraide replied:
Yeah, I noticed, but imo the QOI is worse if you disallow it (because you get a lot of extra diagnostics and it essentially ‘just works’ in codegen if you allow it)

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 10: Expansion Limit) (#169689)

This implements a limit on the maximum number of expansions; this isn’t
in the standard (there is nothing on this in [implimits]), but my
reasoning is that if someone makes a mistake and inadvertently causes
e.g. 1 000 000 000 instantiations, Clang is going to hang... basically
for ever (for comparison, 1 million instantiations takes about 3 seconds
provided the body of the expansion statement is empty).

For now, I’ve set the limit to `256`, but that’s probably too low. I
don’t have strong opinions on the exact value; I definitely *do* think
that there should be *a* limit though. We *could* in theory reuse the
constexpr-steps value, but it feels a bit weird because this isn’t
really doing constant evaluation so much as just template instantiation.
